### ISSUE TITLE
PT-P900 series support?

### ISSUE THREAD

**Issue by TRSx80** [trialinfo/ptouch-driver#3]
I stumbled across your comment over in the [Brother support Issue in lprint](https://github.com/michaelrsweet/lprint/issues/15#issuecomment-618513079):

> I've added support for PT-P900W [...] to ptouch-driver

And it was like a breath of fresh air!  I have been researching label printers for days, and I really like the idea of the Brother PT-P900 series (wide labels up to 36mm, etc.) but there is very little information on the Internet about them, especially as regards Linux support.  And I can't really afford to "take a guess" on a $300-400 label printer and be "wrong."

So I was wondering if you would indulge me and confirm what works, what does not?  I was thinking about getting the (actually most expensive) ethernet networked version (PT-P950NW)...

**Comment by andreas-gruenbacher** [comment]
Printing to USB-attached printers works pretty well. There's no support for dithering (for photographs and similar), printing to multiple stripes (for example, extra large text across two 36mm labels), or for network printing in ptouch-driver.

In addition, CUPS support for roll fed media is very poor and it's not possible in applications to just select the tape width; the predefined "page" sizes all define a height as well. So you have to use custom page sizes and get the application to print exactly that page without scaling up or down to one of the predefined page sizes. It takes a bit of fiddling in the printer settings to get applications to do that, and each application does this slightly differently.

The network / bluetooth enabled Ptouch printers can also be controlled from a Brother smartphone app, so for one-off labels, that may sometimes be the easiest option.

**Comment by TRSx80** [comment]
Thanks for the very fast reply!

I have been trying to get my head around all the various moving parts and implementations (new and old).  LPrint is a very welcome addition of course, and should (if I am understanding correctly) solve a lot of the network support issues.  Pity the timing, as you correctly pointed out.  However maybe we can all put our heads together and finally solve some of these annoying, long-standing issues (like "page" size).

Michael did state the main reason he wrote the program was to address some changes in CUPS that would affect label printers, so I am not sure if this "page" size issue would also be a part of that or not.  I suppose I would be better to simply ask him, instead of you. :smile: 

Thinking more now though, I guess it's an "application" side problem, isn't it?  Or is it somewhere between the applications and the printer drivers?

Have you ever used gLabels?  If so, do you know how that software handles "continuous" media (if at all)?  Perhaps we should better open a separate issue to discuss this "page" size thing, either here or over there.

At any rate, it sounds as if I could get most of what I need working right now, at least over USB directly.  And it also seems like "the time is right" as far as development.  You seem to be active, and now none other than Michael Sweet himself also appears to be on the case for Brother support with LPrint...  Putting that all together it seems that things will be very soon improving (and already have, in the case of `ptouch-driver` here) for those of us with Brother devices.

And I think I can live without (for foreseeable future) any of those advanced features you mention.

So, maybe I _should_ go ahead and pull the trigger, spend the money and help do my little part in helping hack together what remains (network support, "page" sizes, etc...).  Can you hear me talking myself into it? :smile: 

One thing I have been wondering, how wide is the actual print head on the PT-P900 series?  Or, put differently, how big are the actual margins?  I struggled to find current, accurate information on this point.

**Comment by andreas-gruenbacher** [comment]
> I have been trying to get my head around all the various moving parts and implementations (new and old). LPrint is a very welcome addition of course, and should (if I am understanding correctly) solve a lot of the network support issues.

Some of the modern Brother printers support IPP directly, supposedly including PT-P900W and QL-820NWB. I can't try that out on the PT-P900W because I cannot integrate it into my network (no Ethernet jack and no way to configure the wireless other than by WPS, which my network gear doesn't allow). When I make the QL-820NWB printer available over the network, it becomes visible, but I've been unsuccessful in getting it to do anything or in querying the printer's properties with `ipptool`.

> Have you ever used gLabels? If so, do you know how that software handles "continuous" media (if at all)? Perhaps we should better open a separate issue to discuss this "page" size thing, either here or over there.

From the gLabels manual: _The _gLabels_ application is a lightweight program for creating labels and business cards for the GNOME desktop environment. It is designed to work with various laser/ink-jet peel-off label and business card sheets that you'll find at most office supply stores._

So gLabels does nothing for actual label printers. 

> One thing I have been wondering, how wide is the actual print head on the PT-P900 series? Or, put differently, how big are the actual margins? I struggled to find current, accurate information on this point.

The [raster command manuals](https://www.brother.co.jp/eng/dev/command/reference/index.aspx) say that the print head on the P900 series printers is 560 pins wide at 360 dpi, wider than 36 mm tapes, but I don't know of a way of using [bleeding](https://en.wikipedia.org/wiki/Bleed_(printing)), so the actual margins you'll get will be nonzero. It's possible to ignore the 1 mm margin in the feed direction defined in the specification, but the cutter then won't be able to reliably cut between labels.

(The QL printers _do_ have enforced margins.)

**Comment by TRSx80** [comment]
> I can't try that out on the PT-P900W because I cannot integrate it into my network

Yes, I had read your comments over at LPrint about that.

Near as I could tell (by digging around Internet) I don't think it's actually supported on 900 series, but it does seem to be on 750.  For sure the PT-P750W is listed on [Apple AirPrint Supported Devices list](https://support.apple.com/en-us/HT201311) and I am pretty sure I also found "IPP" mentioned specifically in some official Brother PDF I had read somewhere.  I could not find any such information for any 900 series, and I was looking pretty hard for it...  I can be wrong of course...

I did not look for QL-820NWB.

> gLabels

I read the same information at their website.

A little more digging today however, and I found there is apparently a re-qrite using QT that will become gLabels v4 in the [glabels-qt](https://github.com/jimevins/glabels-qt) project, which explicitly states on their README:
> Support for continuous-roll labels

Digging further, I found [this commit](https://github.com/jimevins/glabels-qt/commit/467ca9fc624e07442d45b0214f2cccb4919004a4) which seems to show this feature has been implemented since Aug 11, 2018(!).

> margins

Those seem certainly a lot less than some stuff I have been reading.  But I think that was mostly old information like from Arne Glenstrup's original notes.

In short, this is looking better and better all the time.  I think I am going to sleep on it, but maybe I pull the trigger on one of those here pretty soon.

And it only took like 3 full days of research on my part to get to this point!  :/

Any other thoughts?  How do you like your PT-P900W (if/when you are even able to use it)?

**Comment by andreas-gruenbacher** [comment]
> > gLabels
> A little more digging today however, and I found there is apparently a re-qrite using QT that will become gLabels v4 in the [glabels-qt](https://github.com/jimevins/glabels-qt) project, which explicitly states on their README:
> 
> > Support for continuous-roll labels
> 
> Digging further, I found [this commit](https://github.com/jimevins/glabels-qt/commit/467ca9fc624e07442d45b0214f2cccb4919004a4) which seems to show this feature has been implemented since Aug 11, 2018(!).

That's interesting, I didn't explore that before. The current version of glabels-qt has DK-2205 labels predefined, which are 62 mm continuous rolls. With those, the label length can be entered separately. I've created a test label that printed just fine on the QL-820NWB, with ptouch-driver as the printer driver. So that may be a reasonable way for producing fixed-length labels.

Printers like the P900 series have much narrower tapes, and with those, a more common way of designing labels is to choose a font size and the number of lines, and let the editor determine the length of the label depending on the text, for example. This approach doesn't seem to be supported by glabels-qt, and I also didn't find a way of changing the label length graphically. Those are all things that could be added at some point, of course.

The application I'm working on only requires fixed-length labels so far. For those, I'm using pdf forms that I'm filling out using a simple script on top of poppler. This won't work for variable-length labels, barcodes, etc., but it's good enough for now.

**Comment by andreas-gruenbacher** [comment]
> How do you like your PT-P900W (if/when you are even able to use it)?

I've mostly bought it for making sure ptouch-driver works on current PT printers. The 360 dpi resolution produces much cleaner looking labels than the 180 dpi of older / cheaper models. Like the other Brother PT and QL printers I've used so far, if you feed it reasonable input, the PT-P900W will just work.

**Comment by TRSx80** [comment]
> I've mostly bought it for making sure ptouch-driver works on current PT printers.

Good on you! :beers:

I try to contribute as much support (time, effort, money) to the advancement of F/LOSS as I possibly can, but even I am struggling with the price of this unit (PT-P950NW).

Yes, I am really, really debating myself hard whether to buy the PT-P950NW or not.  They are quite a lot (~$300+) more than 700 series and earlier models.  The main differences (that I can tell) seem to be:
* printing up to 36mm (instead of 24, or even 12, depending on model)
* higher DPI (360, as you point out, instead of 180)
* wired network (Ethernet) connectivity(?)
* ~2x faster printing speed (60 vs 30 mm/s at highest DPI) (not that important to me, personally)

Usually, I don't mind spending a little extra for something a little nicer, which may also include use-cases I do not (currently) have in mind.  Historically this has been a good approach for me.  And yet I am really struggling right now to reconcile a couple potential use-cases[1] with the cost of the printer being 300-400 % (!) the cost of printers without those couple options.  That is a very steep cost increase, for just a couple more features.

Note [1]: The only ones I can come up with are:
* being able to print 3 (or more?) lines of text on one label
* like above, but including graphics, barcodes maybe (more height), even 2D barcodes (QR, etc.)
* being able to use over wired network (Ethernet)

Writing this out now has somewhat clarified my thoughts.  Unless I am missing something with regard to wired Ethernet being available on cheaper models, this feature alone might be worth the (unfortunately steep) increase in price.  Taking the connectivity struggles you have had to deal with into account would only seems to strengthen this point.

**Comment by andreas-gruenbacher** [comment]
PT-P700s are around €50. My PT-P900W was around €180. PT-P950NWs are still around €400.
(Note that the P900 series also support ESC/P, which may explain some of the price gap from the PT-P700.)

**Comment by philpem** [comment]
> So gLabels does nothing for actual label printers.

That's not strictly correct. gLabels works fine with the plastic-tape P-Touch printers and the QL series, using fixed size label stock. (Sample size - my PT-2450DX, QL-500 and QL-570).

The PT series technically is a continuous-stock label printer, you set the page size as Custom in the CUPS driver. QL series I expect would work similarly.

The printers expect data in "portrait" orientation, the edge which comes out of the printer first is the "top" while gLabels works in "landscape" by default. Switching on the 'rotated label stock' option is enough to fix this.

I've been using gLabels with these printers for years but I had to add the label stock to it as a custom page size.

**Comment by TRSx80** [comment]
I just couldn't justify spending that kind of money on a goddamn label printer, I'm sorry but the price that Brother wants for those PT-P950NW is just outrageous.  I ended up scoring a good deal on an old PT-9500PC on eBay, which I guess is an older version (2 versions older?) of basically the same thing.

Then I go shopping for tape...  Well, we all know tape is outrageous, too.  Did you guys know however that you can buy tape direct from the same third party manufacturers who are top sellers on Amazon for like $2-3 per roll, shipped, in single quantities on AliExpress?!  I just wanted to share that with you, as I was astounded.  It takes a month or two to get here of course...  Because of the long lead time (and low price!), I might have went a little nuts ordering tape...  :laughing:

Cheers!  :beers:
